cpp/src/io/utilities/thread_pool.hpp

```diff
@@ -0,0 +1,401 @@
+   */
+  template <typename T, typename F>
+  void parallelize_loop(T first_index, T last_index, const F &loop, ui32 num_tasks = 0)
+  {
+    if (num_tasks == 0) num_tasks = thread_count;
+    if (last_index < first_index) std::swap(last_index, first_index);
+    size_t total_size = last_index - first_index + 1;
+    size_t block_size = total_size / num_tasks;
+    if (block_size == 0) {
+      block_size = 1;
+      num_tasks  = (ui32)total_size > 1 ? (ui32)total_size : 1;
+    }
+    std::atomic<ui32> blocks_running = 0;
+    for (ui32 t = 0; t < num_tasks; t++) {
+      T start = (T)(t * block_size + first_index);
+      T end   = (t == num_tasks - 1) ? last_index : (T)((t + 1) * block_size + first_index - 1);
+      blocks_running++;
+      push_task([start, end, &loop, &blocks_running] {
+        for (T i = start; i <= end; i++) loop(i);
+        blocks_running--;
+      });
+    }
+    while (blocks_running != 0) { sleep_or_yield(); }
+  }
+
+  /**
+   * @brief Push a function with no arguments or return value into the task queue.
+   *
+   * @tparam F The type of the function.
+   * @param task The function to push.
+   */
+  template <typename F>
+  void push_task(const F &task)
+  {
+    tasks_total++;
+    {
+      const std::scoped_lock lock(queue_mutex);
+      tasks.push(std::function<void()>(task));
+    }
+  }
+
+  /**
+   * @brief Push a function with arguments, but no return value, into the task queue.
+   * @details The function is wrapped inside a lambda in order to hide the arguments, as the tasks
+   * in the queue must be of type std::function<void()>, so they cannot have any arguments or return
+   * value. If no arguments are provided, the other overload will be used, in order to avoid the
+   * (slight) overhead of using a lambda.
+   *
+   * @tparam F The type of the function.
+   * @tparam A The types of the arguments.
+   * @param task The function to push.
+   * @param args The arguments to pass to the function.
+   */
+  template <typename F, typename... A>
+  void push_task(const F &task, const A &... args)
+  {
+    push_task([task, args...] { task(args...); });
+  }
+
+  /**
+   * @brief Reset the number of threads in the pool. Waits for all currently running tasks to be
+   * completed, then destroys all threads in the pool and creates a new thread pool with the new
+   * number of threads. Any tasks that were waiting in the queue before the pool was reset will then
+   * be executed by the new threads. If the pool was paused before resetting it, the new pool will
+   * be paused as well.
+   *
+   * @param _thread_count The number of threads to use. The default value is the total number of
+   * hardware threads available, as reported by the implementation. With a hyperthreaded CPU, this
+   * will be twice the number of CPU cores. If the argument is zero, the default value will be used
+   * instead.
+   */
+  void reset(const ui32 &_thread_count = std::thread::hardware_concurrency())
```

Comment:
By name alone `reset` implies that the queue is dumped to me. It seems this function is more `set_thread_count`